### PR TITLE
Display progress during link recovery

### PR DESCRIPTION
**Pull request by jsteemann**
### Scope & Purpose

Displays progress report during arangosearch link recovery, e.g.
```
...
2022-09-05T18:02:00Z [65850] INFO [d1f18] {arangosearch} recovering arangosearch link 4792073 (stage1, operation 1/1 (0%)
2022-09-05T18:03:00Z [65850] INFO [d1f18] {arangosearch} recovering arangosearch link 4792073 (stage3, operation 67/209 (31%)
2022-09-05T18:04:01Z [65850] INFO [d1f18] {arangosearch} recovering arangosearch link 4792073 (stage3, operation 136/209 (64%)
2022-09-05T18:05:01Z [65850] INFO [d1f18] {arangosearch} recovering arangosearch link 4792073 (stage3, operation 208/209 (99%)
2022-09-05T18:05:03Z [65850] INFO [d1f18] {arangosearch} recovering arangosearch link 5507397 (stage1, operation 1/1 (0%)
2022-09-05T18:06:04Z [65850] INFO [d1f18] {arangosearch} recovering arangosearch link 5507397 (stage3, operation 68/209 (32%)
2022-09-05T18:07:05Z [65850] INFO [d1f18] {arangosearch} recovering arangosearch link 5507397 (stage3, operation 137/209 (65%)
2022-09-05T18:08:05Z [65850] INFO [d1f18] {arangosearch} recovering arangosearch link 5507397 (stage3, operation 209/209 (99%)
2022-09-05T18:08:07Z [65850] INFO [d1f18] {arangosearch} recovering arangosearch link 5507407 (stage1, operation 1/1 (0%)
2022-09-05T18:09:07Z [65850] INFO [d1f18] {arangosearch} recovering arangosearch link 5507407 (stage3, operation 63/210 (29%)
2022-09-05T18:10:08Z [65850] INFO [d1f18] {arangosearch} recovering arangosearch link 5507407 (stage3, operation 123/210 (58%)
2022-09-05T18:11:08Z [65850] INFO [d1f18] {arangosearch} recovering arangosearch link 5507407 (stage3, operation 192/210 (90%)
```

- [ ] :hankey: Bugfix
- [x] :pizza: New feature
- [ ] :fire: Performance improvement
- [ ] :hammer: Refactoring/simplification

### Checklist

- [ ] Tests
  - [ ] **Regression tests**
  - [ ] C++ **Unit tests**
  - [ ] **integration tests**
  - [ ] **resilience tests**
- [ ] :book: CHANGELOG entry made
- [ ] :books: documentation written (release notes, API changes, ...)
- [x] Backports
  - [x] Backport for 3.10: https://github.com/arangodb/arangodb/pull/17071
  - [ ] Backport for 3.9: -
  - [ ] Backport for 3.8: -

#### Related Information

- [ ] Docs PR: 
- [ ] Enterprise PR:
- [ ] GitHub issue / Jira ticket:
- [ ] Design document:

### CODE REVIEW

**Comment by Simran-B** [comment]
> operation 1/1 (0%)

Shouldn't that be 100%? Or is op 1 ongoing?
0/1 (0%) would be more intuitive in that case.

**Comment by jsteemann** [comment]
> > operation 1/1 (0%)
> 
> Shouldn't that be 100%? Or is op 1 ongoing? 0/1 (0%) would be more intuitive in that case.

It is operation 1 of 1 ongoing.

**Comment by Simran-B** [comment]
👍
It's a bit misleading that the percentage shows the degree of completion but the op fraction does not

File: arangod/IResearch/IResearchFeature.cpp [hunk 1]
```diff
@@ -1142,6 +1142,27 @@ void IResearchFeature::unprepare() {
...
+                                              std::string_view phase,
+                                              size_t current, size_t total) {
+  TRI_ASSERT(total != 0);
+  auto now = std::chrono::system_clock::now();
+
+  if (id != _progressState.lastReportId ||
+      now - _progressState.lastReportTime >= std::chrono::minutes(1)) {
+    // report progress only when link id changes or one minute has passed
+
+    auto progress = static_cast<int>(100.0 * current / total);
```
Comment (gnusi) [inline arangod/IResearch/IResearchFeature.cpp:1155]: `static_cast<size_t>`

**Review by gnusi — APPROVED** [review]

File: arangod/IResearch/IResearchFeature.cpp [hunk 2]
```diff
@@ -1142,6 +1142,27 @@ void IResearchFeature::unprepare() {
...
+  TRI_ASSERT(total != 0);
+  auto now = std::chrono::system_clock::now();
+
+  if (id != _progressState.lastReportId ||
+      now - _progressState.lastReportTime >= std::chrono::minutes(1)) {
+    // report progress only when link id changes or one minute has passed
+
+    auto progress = static_cast<size_t>(100.0 * current / total);
+    LOG_TOPIC("d1f18", INFO, TOPIC)
+        << "recovering arangosearch link " << id << ", " << phase
```
Comment (Dronplane) [inline arangod/IResearch/IResearchFeature.cpp:1157]: Here we may have link or inverted index, Maybe use more "generic" entity name 
```suggestion
        << "recovering arangosearch index " << id << ", " << phase
```

Comment (jsteemann) [reply arangod/IResearch/IResearchFeature.cpp]: Done

**Pull request merged**